jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

```diff
@@ -1925,50 +1842,53 @@ public void setAsyncAttributes()
         if (getAttribute(AsyncContext.ASYNC_REQUEST_URI) != null)
             return;
 
-        String requestURI;
-        String contextPath;
-        String servletPath;
-        String pathInfo;
-        String queryString;
-        HttpServletMapping httpServletMapping;
-
-        // Have we been forwarded before?
-        requestURI = (String)getAttribute(RequestDispatcher.FORWARD_REQUEST_URI);
-        if (requestURI != null)
-        {
-            contextPath = (String)getAttribute(RequestDispatcher.FORWARD_CONTEXT_PATH);
-            servletPath = (String)getAttribute(RequestDispatcher.FORWARD_SERVLET_PATH);
-            pathInfo = (String)getAttribute(RequestDispatcher.FORWARD_PATH_INFO);
-            queryString = (String)getAttribute(RequestDispatcher.FORWARD_QUERY_STRING);
-            httpServletMapping = (HttpServletMapping)getAttribute(RequestDispatcher.FORWARD_MAPPING);
-        }
-        else
-        {
-            requestURI = getRequestURI();
-            contextPath = getContextPath();
-            servletPath = getServletPath();
-            pathInfo = getPathInfo();
-            queryString = getQueryString();
-            httpServletMapping = getHttpServletMapping();
-        }
-
         // Unwrap the _attributes to get the base attributes instance.
         Attributes baseAttributes;
         if (_attributes == null)
```

Comment:
This whole section really needs some comments - the pattern of handling the AsyncAttributes appears to be very different to handling the Forward/IncludeAttributes that we should explain why.

Reply:
extra comments in cf0a795475d78d27743bf331b5134a85515c45e7